### ISSUE TITLE
dockerTools.buildImage deletes base image's /bin directory

### ISSUE THREAD

**Issue by vroad** [NixOS/nixpkgs#110696]
**Describe the bug**
If centos image is specified for fromImage argument of dockerTools.buildImage, everything in /bin disappears (except packages installed with nix).

**Steps To Reproduce**

1. clone my repository: https://github.com/vroad/nix-docker-base-image-test
2. nix-build docker.nix
3. docker load -i result
4. Run `docker run --rm -it vroad/docker-base-image-test:latest ls -lah /bin/` to see files from base image is still there.

**Expected behavior**

/bin retains contents from centos image.

**`nix-env --version` output**
nix-env (Nix) 2.3.10

**Additional context**

If I specify empty array to `contents` argument, /bin is retained.

I can't run any RUN instructions on Dockerfile with the newly created docker image as base, because '/bin/sh' no longer exist.

**Comment by zimbatm** [comment]
Transferred to the right repo.

I think that the original /bin files are still being pulled and shadowed by the nix layers.

**Comment by vroad** [comment]
Thank you. /bin appears to be a symlink to some other directory, at least on centos.
So, is this expected behavior? Can I install binaries to somewhere else to prevent /bin from being overwritten by nix-build?

**Comment by zimbatm** [comment]
I don't think that the builder really has the knowledge of the previous layers. If a symlink gets replaced by a folder, it's going to shadow it.

Fedora and friends have unified /bin and /usr/bin so it's probably possible to re-symlink those binaries into /bin using the `runAsRoot` instruction. It would look something like this:

```nix
buildImage {
  # ...
  runAsRoot = ''
    #!{pkgs.stdenv.shell}
    set -euo pipefail
    for file in usr/bin/* ; do
      ln -s "/$file" "bin/$(basename "$file")"
    done
  '';
```

**Comment by hhoeflin** [comment]
Hi,

I am running into a similar problem. As I don't have sudo rights on the machine where I want to build these, I wanted to avoid the "runAsRoot" part. It would be very helpful if it was possible to tell the buildImage to link the binaries to a different location other than "/bin". Is this an option?

Thanks

**Comment by sanmai-NL** [comment]
@hhoeflin 
> runAsRoot is a bash script that will run as root in an environment that overlays the existing layers of the base image with the new resulting layer, including the previously copied contents derivation. This can be similarly seen as RUN ... in a Dockerfile.

`runAsRoot` does not run in the context of your builder host, but in the temporary build container itself. So you can run things as root freely.
